### PR TITLE
Fix udp datagram socket

### PR DESCRIPTION
**Pull request by jakobod**
This PR fixes the `udp_datagram_socket` test. It also fixes a bug in `convert_ip_endpoint` and changes lookup of the local address to us the new `local_addresses()` function.

### CODE REVIEW

File: libcaf_net/src/udp_datagram_socket.cpp [hunk 1]
```diff
@@ -100,9 +100,8 @@ variant<std::pair<size_t, ip_endpoint>, sec> read(udp_datagram_socket x,
                          << CAF_ARG(buf.size()) << " of " << CAF_ARG(num_bytes)
                          << " bytes");
     ip_endpoint ep;
-    // TODO: how to properly pass error further?
     if (auto err = detail::convert(addr, ep))
-      return sec::runtime_error;
```
Comment (Neverlord) [inline libcaf_net/src/udp_datagram_socket.cpp]: I think we should add `CAF_ASSERT(err.category == atom("system"))` here to make sure the `static_cast` below is safe.

File: libcaf_net/test/udp_datagram_socket.cpp [hunk 2]
```diff
@@ -30,85 +30,68 @@
...
-  if (contains(v4_local)) {
-    test_send_receive(v4_local);
-  } else if (contains(v6_local)) {
-    test_send_receive(v6_local);
-  } else {
-    CAF_FAIL("could not resolve 'localhost'");
-  }
+  std::vector<byte> buf(1024);
+  if (auto err = nonblocking(socket_cast<net::socket>(receive_socket), true))
+    CAF_FAIL("nonblocking returned an error" << err);
```
Comment (Neverlord) [inline libcaf_net/test/udp_datagram_socket.cpp]:
```suggestion
    CAF_FAIL("setting socket to nonblocking failed: " << err);
```

File: libcaf_net/test/udp_datagram_socket.cpp [hunk 3]
```diff
@@ -30,85 +30,68 @@
...
+  variant<std::pair<size_t, ip_endpoint>, sec> read_ret;
+  do {
+    read_ret = read(receive_socket, make_span(buf));
+    if (auto read_res = get_if<std::pair<size_t, ip_endpoint>>(&read_ret)) {
+      CAF_CHECK_EQUAL(read_res->first, hello_test.size());
+      buf.resize(read_res->first);
+    } else if (get<sec>(read_ret) != sec::unavailable_or_would_block) {
+      CAF_FAIL("read returned an error: " << sys.render(get<sec>(read_ret)));
+    }
+    CAF_CHECK_LESS(++rounds, 100);
```
Comment (Neverlord) [inline libcaf_net/test/udp_datagram_socket.cpp]:
```suggestion
    if (++rounds == 100)
      CAF_FAIL("unable to receive data from the UDP socket");
```

File: libcaf_net/test/udp_datagram_socket.cpp [hunk 4]
```diff
@@ -30,85 +30,68 @@
...
+  } while (holds_alternative<sec>(read_ret)
```
Comment (Neverlord) [inline libcaf_net/test/udp_datagram_socket.cpp]: We call `CAF_FAIL` for any other error reason. What we actually want here is to wait until we have a valid result, I think this version is easier to comprehend:

```suggestion
  } while (read_ret.index() != 0);
```

File: libcaf_net/test/udp_datagram_socket.cpp [hunk 5]
```diff
@@ -30,85 +30,68 @@
...
-  }
+  std::vector<byte> buf(1024);
+  if (auto err = nonblocking(socket_cast<net::socket>(receive_socket), true))
+    CAF_FAIL("nonblocking returned an error" << err);
+  CAF_CHECK_EQUAL(read(receive_socket, make_span(buf)),
+                  sec::unavailable_or_would_block);
+  CAF_MESSAGE("sending data to " << to_string(ep));
+  CAF_CHECK_EQUAL(write(send_socket, as_bytes(make_span(hello_test)), ep),
+                  hello_test.size());
+  int rounds = 0;
```
Comment (Neverlord) [inline libcaf_net/test/udp_datagram_socket.cpp]: It's unclear from the context what a "round" is supposed to be here IMO. I think renaming the variable makes the intent clearer:

```suggestion
  int receive_attempts = 0;
```

File: libcaf_net/test/udp_datagram_socket.cpp (hunk 4 above)
Comment (jakobod) [reply libcaf_net/test/udp_datagram_socket.cpp]: What exactly does `index()` do?

Comment (Neverlord) [reply libcaf_net/test/udp_datagram_socket.cpp]: It asks the `variant` [which type it contains](https://en.cppreference.com/w/cpp/utility/variant/index). Here, 0 is the index of the pair. Using `holds_alternative` is also OK, but only if we give the pair some typedef to reduce verbosity. :)

**Pull request merged**